File: operations/src/main/java/l/files/operations/Move.java
package l.files.operations;

import java.io.IOException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.concurrent.atomic.AtomicInteger;

import static java.nio.file.Files.move;

final class Move extends Paste {

    private final AtomicInteger movedItemCount = new AtomicInteger();

    Move(Collection<? extends Path> sourcePaths, Path destinationPath) {
        super(sourcePaths, destinationPath);
    }

    int getMovedItemCount() {
        return movedItemCount.get();
    }

    @Override
    void paste(Path sourcePath, Path destinationPath) throws IOException {
        try {
            move(sourcePath, destinationPath);
            movedItemCount.incrementAndGet();
        } catch (IOException e) {
            record(sourcePath, e);
        }
    }

}

File: files-app/src/main/kotlin/operations/Move.kt
package l.files.operations

import java.io.IOException

import l.files.fs.Path

private class Move(src: Iterable<Path>, dst: Path) : Paste(src, dst) {

    volatile var movedItemCount = 0
        private set

    override fun paste(from: Path, to: Path, listener: FailureRecorder) {
        try {
            from.resource.move(to)
            movedItemCount++
        } catch (e: IOException) {
            listener.onFailure(from, e)
        }
    }

}
